### ISSUE TITLE
Handling unsuccessful documents upsert / delete via Bulk API

### ISSUE THREAD

**Issue by cieciurm** [Azure/azure-cosmos-dotnet-v3#4748]
Hi,

I've migrated my application using `Azure CosmosDB SDK` to v3 and I'm using the bulk functionality to upsert somewhere between 200 and 1000 items.

Using the `BulkExecutor` and its configuration I was getting 100% success rate when upserting - all documents were saved.

In V3 I'm getting 429 response, and only some documents are saved. I'm not sure how I should handle those that were not saved.

`ClientOptions` setup:

```csharp
var options = new CosmosDbOptions
{
  AllowBulkExecution = true,
  MaxRetryWaitTimeOnRateLimitedRequests = TimeSpan.FromSeconds(60),
  MaxRetryAttemptsOnRateLimitedRequests = 19,
}
```

I'm using the `BulkOperations<T>` wrapper as described in the [docs](https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/how-to-migrate-from-bulk-executor-library).

My main method for bulk upserting looks like this:

```csharp
public async Task<BulkOperationResponse<TDocument>> ImportAsync(IEnumerable<TDocument> documents)
{
     var bulkOperations = new BulkOperations<TDocument>(documents.Count());

    foreach (var document in documents)
    {
        bulkOperations.Tasks.Add(CaptureOperationResponse(_container.Value.CreateItemAsync(document, new PartitionKey(document.PartitionKey)), document));
    }

    var response = await bulkOperations.ExecuteAsync();
    return response;
}
```

After inspecting the `BulkOperationResponse<T>` I often see that only a chunk of documents, were saved.

![image](https://github.com/user-attachments/assets/cdae5b2a-66a1-44be-acea-4c48b952a48f)

I have the same issue when trying to bulk delete documents via a method:

```csharp
public async Task<BulkOperationResponse<TDocument>> DeleteAsync(IEnumerable<TDocument> documents)
{
    var bulkOperations = new BulkOperations<TDocument>(documents.Count());

    foreach (var document in documents)
    {
        bulkOperations.Tasks.Add(CaptureOperationResponse(_container.Value.DeleteItemAsync<TDocument>(document.Id, new PartitionKey(document.PartitionKey)), document));
    }

    var response = await bulkOperations.ExecuteAsync();

    return response;
}
```

I'm using a shared throughput for 4 containers - 400 RU/s.

What's the recommended approach to retry upserting of the failed documents?
Should it be handled by the SDK or should I retry in my code?

Thanks in advance for any guidance :)

**Environment summary**
SDK Version: 3.43.1

**Comment by philipthomas-MSFT** [comment]
> I'm using a shared throughput for 4 containers - 400 RU/s.
> What's the recommended approach to retry upserting of the failed documents?
> Should it be handled by the SDK or should I retry in my code?

Hello @cieciurm,

Thank you for your patience, and for reaching out to us.

While the Cosmos DB SDK has built-in retry logic—which you're already configuring using `MaxRetryAttemptsOnRateLimitedRequests` and `MaxRetryWaitTimeOnRateLimitedRequests`—it’s also a good practice to implement retries at the application level. This can give you more control over how failures are handled, especially in scenarios with high throughput or when frequent 429 (Too Many Requests) errors occur.

You might want to introduce an explicit retry mechanism with an **exponential backoff** strategy. This approach increases the delay between each retry attempt after receiving a 429, reducing the likelihood of overloading the system again.

Implementing retries at the application level provides an extra layer of resilience, ensuring that your operations succeed even when the built-in retry logic has exhausted its limits.

Let me know if you have any further questions!

**Comment by philipthomas-MSFT** [comment]
Hello @cieciurm,

Please feel free to reopen this issue if it has not yet been fully mitigated or resolved. Our priority is to ensure that your needs are thoroughly addressed and that you have a seamless and pleasant experience while using our product.

Do not hesitate to reach out with any further questions, feedback, or suggestions. We greatly value your input and are always here to assist you.

Thank you for your continued trust and support.

**Comment by cieciurm** [comment]
Hey @philipthomas-MSFT,

Thanks for your answer. I analyzed how the code behaves in previous version vs. V3

I noticed that upserting 100 documents in `BulkExecutor` took over 2 seconds and consumed over 2 000 RU.

![Image](https://github.com/user-attachments/assets/983da687-70e7-4775-a5d5-71bc6a0c5633)

In `V3` it took 7 seconds, consumed almost 2 000 RU, while some of the documents failed.

![Image](https://github.com/user-attachments/assets/4d82d855-256b-49ac-a4e4-d98efe5032ce)

So I decided to upscale before my code to a significantly larger throughput and downscale after the operation finishes. This gave me the desired results.

![Image](https://github.com/user-attachments/assets/a31ffc1e-4688-4451-b806-263bc44ffed3)

I did some reading up and there are more ways to handle it, such as:

- switch from manual throughput to auto-scale
- switch to serverless throughput

But due to my setup, I decided to stick with manually increasing the throughput for the time of my operation.

I'm just wondering how come `BulkExecutor` code didn't get throttled... It was also exceeding the 400 RU/s threshold that I've setup for the database... It seems to have consumed ~1 000 RU/s, yet all of the documents were saved correctly.

Any clue how's that possible?

Thanks!
